### PR TITLE
Only pass extra create_trail options if set in module params

### LINKED ISSUE #34700: cloudtrail module does not support China region
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
cloudtrail

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.5.0 (devel b88304f211) last updated 2018/01/10 15:31:29 (GMT +100)
  config file = /Users/nbailey/automation/roles-aws/aws_cloudtrail/tests/ansible.cfg
  configured module search path = [u'/Users/nbailey/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/nbailey/virtualenv-ansible-devel/src/ansible/lib/ansible
  executable location = /Users/nbailey/virtualenv-ansible-devel/bin/ansible
  python version = 2.7.10 (default, Oct 23 2015, 19:19:21) [GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)]
```

##### CONFIGURATION
nothing relevant.

##### OS / ENVIRONMENT

OSX, running against AWS China.

##### SUMMARY

Boto3 itself fully supports CloudTrail in the AWS China region. Ansible's implementation does not seem to - CreateTrail API calls return with an HTTP 400 and the following: 

"An error occurred (UnsupportedOperationException) when calling the CreateTrail operation: The operation requested is not supported in the region."

I can't see anything to cause this. Behaviour is the same when explicitly setting the documented China Cloudtrail endpoint (https://cloudtrail.cn-north-1.amazonaws.com.cn) and when allowing Boto3 to evaluate endpoints itself.

##### STEPS TO REPRODUCE

```yaml
- name: create test cloudtrail in china
  hosts: test-trail-1
  connection: local
  tags: [ china ]

  pre_tasks:
  - name: CreateTrail API call
    cloudtrail:
      state: enabled
      name: test-trail-1
      s3_bucket_name: "{{ aws.bucket_name }}"
      region: cn-north-1
      profile: "{{ aws.boto_profile }}"
```

##### EXPECTED RESULTS
The module creates a trail, just as they do when creating a boto3 cloudtrail client and making the API calls directly:

```
$ python
Python 2.7.10 (default, Oct 23 2015, 19:19:21)
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import boto3
>>> import botocore
>>> session = boto3.Session(profile_name='<redacted>')
>>> session = boto3.Session(profile_name='<redacted>')
>>> preprod = session.client('cloudtrail', region_name='cn-north-1')
>>> response = preprod.delete_trail(Name='test')
>>> response
{'ResponseMetadata': {'RetryAttempts': 0, 'HTTPStatusCode': 200, 'RequestId': 'b289adf1-3b5e-47bd-b883-d29ee3a6d216', 'HTTPHeaders': {'x-amzn-requestid': 'b289adf1-3b5e-47bd-b883-d29ee3a6d216', 'date': 'Wed, 10 Jan 2018 15:35:20 GMT', 'content-length': '2', 'content-type': 'application/x-amz-json-1.1'}}}
>>> response = preprod.create_trail(Name='test',S3BucketName='cn-sh-cnn1-cloud-trail')
>>> response
{u'IncludeGlobalServiceEvents': True, u'Name': u'test', 'ResponseMetadata': {'RetryAttempts': 0, 'HTTPStatusCode': 200, 'RequestId': '708da52e-e384-403f-aa16-0fa53c1933bd', 'HTTPHeaders': {'x-amzn-requestid': '708da52e-e384-403f-aa16-0fa53c1933bd', 'date': 'Wed, 10 Jan 2018 15:35:31 GMT', 'content-length': '186', 'content-type': 'application/x-amz-json-1.1'}}, u'TrailARN': u'arn:aws-cn:cloudtrail:cn-north-1:640673851126:trail/test', u'IsMultiRegionTrail': False, u'S3BucketName': u'cn-sh-cnn1-cloud-trail'}
>>>
```

##### ACTUAL RESULTS

 The module appears to connect to the AWS service endpoint in cn-north-1 successfully (the region parameter is correct and it is an AWS service response). However the module output reports that the CreateTrail operation is not supported in this region, even though it works with the AWS CLI or with a boto3 API call not wrapped by Ansible.

Here's full output:
 
```
RUNNING HANDLER [aws_cloudtrail : manage cloudtrail trails] ****************************************************************************************************************************************************************************************************************************************
Using module file /Users/nbailey/automation/roles-aws/aws_modules/library/cloudtrail.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: nbailey
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/nbailey/.ansible/tmp/ansible-tmp-1515598328.47-33693836653100 `" && echo ansible-tmp-1515598328.47-33693836653100="` echo /Users/nbailey/.ansible/tmp/ansible-tmp-1515598328.47-33693836653100 `" ) && sleep 0'
<localhost> PUT /var/folders/3k/n0flxzzx3tn15hd44_0l4qt40038gw/T/tmpvORY0v TO /Users/nbailey/.ansible/tmp/ansible-tmp-1515598328.47-33693836653100/cloudtrail.py
<localhost> EXEC /bin/sh -c 'chmod u+x /Users/nbailey/.ansible/tmp/ansible-tmp-1515598328.47-33693836653100/ /Users/nbailey/.ansible/tmp/ansible-tmp-1515598328.47-33693836653100/cloudtrail.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /Users/nbailey/.ansible/tmp/ansible-tmp-1515598328.47-33693836653100/cloudtrail.py; rm -rf "/Users/nbailey/.ansible/tmp/ansible-tmp-1515598328.47-33693836653100/" > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/3k/n0flxzzx3tn15hd44_0l4qt40038gw/T/ansible_MCTZx7/ansible_module_cloudtrail.py", line 270, in create_trail
    resp = client.create_trail(**ct_params)
  File "/Users/nbailey/Library/Python/2.7/lib/python/site-packages/botocore/client.py", line 317, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/nbailey/Library/Python/2.7/lib/python/site-packages/botocore/client.py", line 615, in _make_api_call
    raise error_class(parsed_response, operation_name)
UnsupportedOperationException: An error occurred (UnsupportedOperationException) when calling the CreateTrail operation: The operation requested is not supported in the region.

fatal: [test-trail-1 -> localhost]: FAILED! => {
    "attempts": 1,
    "changed": false,
    "error": {
        "code": "UnsupportedOperationException",
        "message": "The operation requested is not supported in the region."
    },
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "cloudwatch_logs_log_group_arn": null,
            "cloudwatch_logs_role_arn": null,
            "ec2_url": null,
            "enable_log_file_validation": true,
            "enable_logging": false,
            "include_global_events": true,
            "is_multi_region_trail": true,
            "kms_key_id": null,
            "name": "test-trail-1",
            "profile": "<redacted>",
            "region": "cn-north-1",
            "s3_bucket_name": "aws-cloudtrail-test-bucket-a",
            "s3_key_prefix": null,
            "security_token": null,
            "sns_topic_name": null,
            "state": "present",
            "tags": {},
            "validate_certs": true
        }
    },
    "msg": "An error occurred (UnsupportedOperationException) when calling the CreateTrail operation: The operation requested is not supported in the region.",
    "response_metadata": {
        "http_headers": {
            "connection": "close",
            "content-length": "110",
            "content-type": "application/x-amz-json-1.1",
            "date": "Wed, 10 Jan 2018 15:32:10 GMT",
            "x-amzn-requestid": "<redacted>"
        },
        "http_status_code": 400,
        "request_id": "<redacted>",
        "retry_attempts": 0
    }
}
```

Not sure how to debug this further - I'm no great expert on the boto framework, but I don't see anything weird in either the module (create_trail function is here: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/cloudtrail.py#L266
 and fairly straightforward) or in the boto_conn logic (here: https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ec2.py#L116). Any insight into it much appreciated!

### PR DESCRIPTION
**Pull request by kwerey**
##### SUMMARY

Fixes #34700  - the Amazon CloudTrail service in China (and I think some other regions) does not support the EnableLogFileValidation and KmsKeyId parameters.

The module currently defaults to setting the former as false and passing the latter as an empty string, which causes an UnsupportedOperationException.

This PR takes out the empty-string defaults, and lets the CloudTrail API default log file validation to false (see [here](https://docs.aws.amazon.com/awscloudtrail/latest/APIReference/API_CreateTrail.html#awscloudtrail-CreateTrail-request-EnableLogFileValidation) for the API reference).

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

cloudtrail

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.5.0 (cloudtrail-china-support 4913991103) last updated 2018/01/11 17:34:41 (GMT +100)
  config file = /Users/nbailey/.ansible.cfg
  configured module search path = [u'/Users/nbailey/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/nbailey/ansible/lib/ansible
  executable location = /Users/nbailey/ansible/bin/ansible
  python version = 2.7.10 (default, Oct 23 2015, 19:19:21) [GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)]
```

Fix verified in US and China AWS regions; happy to adjust formatting if there are Ansible module-writing conventions I'm missing.

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/cloudtrail.py
Comment (tedder) [inline lib/ansible/modules/cloud/amazon/cloudtrail.py]: Hmm. The problem with this is that it isn't possible to turn off log file validation, making this fail the idempotency test (you can't assert that log file validation is disabled). I'm not sure if the 'bool' param allows for the param to be [omitted](http://docs.ansible.com/ansible/latest/playbooks_filters.html#omitting-parameters) or if it defaults to false; if you can't test that, I'll try to get to it in the next few days/weeks.

**Review by tedder — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/cloudtrail.py
Comment (kwerey) [reply lib/ansible/modules/cloud/amazon/cloudtrail.py]: Huh... I don't think I totally follow, sorry. If you tried to assert it was disabled, anything getting info from 'get-trail-status' would report that it is disabled whether it's set to false or not specified, because "disabled" is the default of the service itself for new trails.

Using default(omit) in a task doesn't remove module defaults, I don't think - just only passes task params if they're set. Happy to test tomorrow UK time.

Comment (tedder) [reply lib/ansible/modules/cloud/amazon/cloudtrail.py]: Here's the use case:

- enable cloudtrail, set log file validation=true
- set log file validation=false

Given the conditional in 488, validation would still be enabled.

Comment (shepdelacreme) [reply lib/ansible/modules/cloud/amazon/cloudtrail.py]: I just tested this and can confirm that once you enable log file validation you can't set it back to False due to the conditional on 488.

`AnsibleModule` by default sets parameters to None if they are not specified and do not have a Default set, regardless of their type. So in the current code I think the above behavior can be avoided by changing line 488 to:
```python
if module.params['enable_log_file_validation'] is not None:
    ct_params['EnableLogFileValidation'] = module.params['enable_log_file_validation']
```

**Comment by shepdelacreme** [comment]
Module fails in check mode due to a KeyError from the code on line  https://github.com/kwerey/ansible/blob/4913991103e5aeb38e609745ca2535daec705134/lib/ansible/modules/cloud/amazon/cloudtrail.py#L598 

```
Traceback (most recent call last):
  File "/var/folders/sk/k6tqwkbj59x4glzs2hzt2vzrbm012h/T/ansible_XcfNj0/ansible_module_cloudtrail.py", line 617, in <module>
    main()
  File "/var/folders/sk/k6tqwkbj59x4glzs2hzt2vzrbm012h/T/ansible_XcfNj0/ansible_module_cloudtrail.py", line 602, in main
    trail.pop('EnableLogFileValidation')
KeyError: 'EnableLogFileValidation'

fatal: [localhost]: FAILED! => {
    "attempts": 1, 
    "changed": false, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/sk/k6tqwkbj59x4glzs2hzt2vzrbm012h/T/ansible_XcfNj0/ansible_module_cloudtrail.py\", line 617, in <module>\n    main()\n  File \"/var/folders/sk/k6tqwkbj59x4glzs2hzt2vzrbm012h/T/ansible_XcfNj0/ansible_module_cloudtrail.py\", line 602, in main\n    trail.pop('EnableLogFileValidation')\nKeyError: 'EnableLogFileValidation'\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 0
}
```

File: lib/ansible/modules/cloud/amazon/cloudtrail.py
Comment (kwerey) [reply lib/ansible/modules/cloud/amazon/cloudtrail.py]: D'oh - yup, that makes perfect sense by the light of day. Will update + fix the check mode.

**Comment by kwerey** [comment]
Updated this PR so it checks the module parameter is not None instead, and updated the check mode mock return values so it sets ct_params['EnableLogFileValidation'] to false if it wasn't set from the module params.

Does this approach make sense?

I'm happy to do something else otherwise, though I'm suspicious of anything that adds checks like `if region = 'cn-north-1'`: there's now a second China region and there'll probably be other similiarly nonstandard service endpoint behaviours in the future...

File: lib/ansible/modules/cloud/amazon/cloudtrail.py
Comment (tedder) [reply lib/ansible/modules/cloud/amazon/cloudtrail.py]: No worries. Thanks for your patience.

File: lib/ansible/modules/cloud/amazon/cloudtrail.py
Comment (tedder) [inline lib/ansible/modules/cloud/amazon/cloudtrail.py:600]: This works in China? I thought the param would have to be missing.

Comment (kwerey) [reply lib/ansible/modules/cloud/amazon/cloudtrail.py:600]: This isn't making a real request to Cloudtrail in China - it's constructing the spoof return values from a request (hence the fake ARN).

Comment (tedder) [reply lib/ansible/modules/cloud/amazon/cloudtrail.py:600]: Aha. I need more coffee. Thanks for doing this.

Comment (kwerey) [reply lib/ansible/modules/cloud/amazon/cloudtrail.py:600]: If I'm reading right, what's happening is that the check mode opens an STS connection a few lines up to make sure you've passed real credentials and it really can connect to the region, but then basically makes up a fake API return output right here... 

```
$ ./hacking/test-module -m lib/ansible/modules/cloud/amazon/cloudtrail.py -a "name='test-trail-1' s3_bucket_name='cn-sh-cnn1-cloud-trail' region='cn-north-1' state=present profile='<redacted>'" -c
* including generated source, if any, saving to: /Users/nbailey/.ansible_module_generated
* ansiballz module detected; extracted module source to: /Users/nbailey/debug_dir
***********************************
RAW OUTPUT

{"invocation": {"module_args": {"profile": "<redacted>", "aws_secret_key": null, "aws_access_key": null, "include_global_events": true, "name": "test-trail-1", "security_token": null, "tags": {}, "validate_certs": true, "region": "cn-north-1", "kms_key_id": null, "cloudwatch_logs_log_group_arn": null, "s3_bucket_name": "cn-sh-cnn1-cloud-trail", "is_multi_region_trail": false, "state": "present", "cloudwatch_logs_role_arn": null, "ec2_url": null, "enable_log_file_validation": null, "sns_topic_name": null, "s3_key_prefix": null, "enable_logging": true}}, "trail": {"name": "test-trail-1", "has_custom_event_selectors": false, "tags": {}, "include_global_service_events": true, "s3_bucket_name": "cn-sh-cnn1-cloud-trail", "home_region": "cn-north-1", "is_logging": true, "is_multi_region_trail": false, "trail_arn": "arn:aws:cloudtrail:cn-north-1:640673851126:trail/test-trail-1"}, "changed": true, "exists": false}

***********************************
PARSED OUTPUT
{
    "changed": true,
    "exists": false,
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "cloudwatch_logs_log_group_arn": null,
            "cloudwatch_logs_role_arn": null,
            "ec2_url": null,
            "enable_log_file_validation": null,
            "enable_logging": true,
            "include_global_events": true,
            "is_multi_region_trail": false,
            "kms_key_id": null,
            "name": "test-trail-1",
            "profile": "<redacted>",
            "region": "cn-north-1",
            "s3_bucket_name": "cn-sh-cnn1-cloud-trail",
            "s3_key_prefix": null,
            "security_token": null,
            "sns_topic_name": null,
            "state": "present",
            "tags": {},
            "validate_certs": true
        }
    },
    "trail": {
        "has_custom_event_selectors": false,
        "home_region": "cn-north-1",
        "include_global_service_events": true,
        "is_logging": true,
        "is_multi_region_trail": false,
        "name": "test-trail-1",
        "s3_bucket_name": "cn-sh-cnn1-cloud-trail",
        "tags": {},
        "trail_arn": "arn:aws:cloudtrail:cn-north-1:640673851126:trail/test-trail-1"
    }
}
```

note: "trail_arn": "arn:aws:cloudtrail:cn-north-1:640673851126:trail/test-trail-1" isn't a possible ARN, the partition name would be `aws-cn` for a real trail.

**Comment by tedder** [comment]
`shipit`

File: lib/ansible/modules/cloud/amazon/cloudtrail.py
Comment (kwerey) [reply lib/ansible/modules/cloud/amazon/cloudtrail.py:600]: Ha, no problem. Thanks for picking up bug reports that affect a tiiiiiiny percentage of Cloudtrail users :)

**Review by s-hertel — APPROVED** [review]

**Comment by s-hertel** [comment]
bot_status

**Comment by kwerey** [comment]
ready_for_review

**Comment by kwerey** [comment]
@s-hertel thanks for poking the bot about this PR. I'd figured since a maintainer marked it as shippable it didn't need any further github workflow wrangling - is anything else needed to move it forward?

**Comment by kwerey** [comment]
Taking a second look - @tedder , my guess is the magical ansibot will wait on you to change your review status from 'requested changes' to 'approved' before merging this in.

**Comment by s-hertel** [comment]
@kwerey Nothing needed, this looks great. Thanks for fixing this. I just wanted to poke the bot to make sure it was safe to merge. :)

**Comment by tedder** [comment]
Thanks s- and k.

**Pull request merged**
